### PR TITLE
Storage tracing

### PR DESCRIPTION
**Pull request by arkpar**
Detailed externalities tracing to help diagnose consensus issues.

### CODE REVIEW

File: core/client/db/src/lib.rs [hunk 1]
```diff
@@ -110,6 +110,12 @@ impl<B: BlockT> Drop for RefTrackingState<B> {
 	}
 }
 
+impl <Block: BlockT> std::fmt::Debug for RefTrackingState<Block> {
```
Comment (pepyakin) [inline core/client/db/src/lib.rs]:
```suggestion
impl<Block: BlockT> std::fmt::Debug for RefTrackingState<Block> {
```

File: core/client/db/src/storage_cache.rs [hunk 2]
```diff
@@ -285,6 +285,12 @@ pub struct CachingState<H: Hasher, S: StateBackend<H>, B: BlockT> {
 	pub cache: CacheChanges<H, B>
 }
 
+impl <H: Hasher, S: StateBackend<H>, B: BlockT> std::fmt::Debug for CachingState<H, S, B> {
```
Comment (pepyakin) [inline core/client/db/src/storage_cache.rs]:
```suggestion
impl<H: Hasher, S: StateBackend<H>, B: BlockT> std::fmt::Debug for CachingState<H, S, B> {
```

File: core/state-machine/Cargo.toml [hunk 3]
```diff
@@ -16,6 +16,7 @@ primitives = { package = "substrate-primitives", path = "../primitives" }
 panic-handler = { package = "substrate-panic-handler", path = "../panic-handler" }
 codec = { package = "parity-scale-codec", version = "1.0.0" }
 num-traits = "0.2"
+rand = "0.5"
```
Comment (bkchr) [inline core/state-machine/Cargo.toml]:
```suggestion
rand = "0.7.1"
```

File: core/primitives/src/hexdisplay.rs [hunk 4]
```diff
@@ -21,28 +21,37 @@ pub struct HexDisplay<'a>(&'a [u8]);
 
 impl<'a> HexDisplay<'a> {
 	/// Create new instance that will display `d` as a hex string when displayed.
-	pub fn from(d: &'a dyn AsBytesRef) -> Self { HexDisplay(d.as_bytes_ref()) }
+	pub fn from<R: AsBytesRef>(d: &'a R) -> Self { HexDisplay(d.as_bytes_ref()) }
 }
 
 impl<'a> ::core::fmt::Display for HexDisplay<'a> {
-	fn fmt(&self, fmtr: &mut ::core::fmt::Formatter) -> Result<(), ::core::fmt::Error> {
+	fn fmt(&self, f: &mut ::core::fmt::Formatter) -> Result<(), ::core::fmt::Error> {
```
Comment (bkchr) [inline core/primitives/src/hexdisplay.rs]:
```suggestion
	fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
```

File: core/primitives/src/hexdisplay.rs [hunk 5]
```diff
@@ -21,28 +21,37 @@ pub struct HexDisplay<'a>(&'a [u8]);
 
 impl<'a> HexDisplay<'a> {
 	/// Create new instance that will display `d` as a hex string when displayed.
-	pub fn from(d: &'a dyn AsBytesRef) -> Self { HexDisplay(d.as_bytes_ref()) }
+	pub fn from<R: AsBytesRef>(d: &'a R) -> Self { HexDisplay(d.as_bytes_ref()) }
```
Comment (bkchr) [inline core/primitives/src/hexdisplay.rs]: Why did you change the signature?

File: core/client/db/src/lib.rs [hunk 6]
```diff
@@ -110,6 +110,12 @@ impl<B: BlockT> Drop for RefTrackingState<B> {
 	}
 }
 
+impl<Block: BlockT> std::fmt::Debug for RefTrackingState<Block> {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		write!(f, "Block {:?}", self.parent_hash)
```
Comment (bkchr) [inline core/client/db/src/lib.rs]: https://doc.rust-lang.org/std/fmt/struct.DebugStruct.html ?

Otherwise you don't know where `Block` belongs to.

File: core/client/db/src/storage_cache.rs [hunk 7]
```diff
@@ -285,6 +285,12 @@ pub struct CachingState<H: Hasher, S: StateBackend<H>, B: BlockT> {
 	pub cache: CacheChanges<H, B>
 }
 
+impl<H: Hasher, S: StateBackend<H>, B: BlockT> std::fmt::Debug for CachingState<H, S, B> {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		write!(f, "Block {:?}", self.cache.parent_hash)
```
Comment (bkchr) [inline core/client/db/src/storage_cache.rs]: Same.

File: core/client/src/light/backend.rs [hunk 8]
```diff
@@ -325,6 +325,15 @@ where
 	}
 }
 
+impl<H: Hasher> std::fmt::Debug for GenesisOrUnavailableState<H> {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match *self {
```
Comment (bkchr) [inline core/client/src/light/backend.rs]: Same.

File: core/primitives/src/hexdisplay.rs [hunk 9]
```diff
@@ -21,28 +21,37 @@ pub struct HexDisplay<'a>(&'a [u8]);
...
 			for byte in &self.0[self.0.len() - 512..] {
-				fmtr.write_fmt(format_args!("{:02x}", byte))?;
+				f.write_fmt(format_args!("{:02x}", byte))?;
 			}
 		}
 		Ok(())
 	}
 }
 
+impl<'a> ::core::fmt::Debug for HexDisplay<'a> {
```
Comment (bkchr) [inline core/primitives/src/hexdisplay.rs]:
```suggestion
impl<'a> core::fmt::Debug for HexDisplay<'a> {
```

File: core/primitives/src/hexdisplay.rs [hunk 10]
```diff
@@ -21,28 +21,37 @@ pub struct HexDisplay<'a>(&'a [u8]);
...
+	fn fmt(&self, f: &mut ::core::fmt::Formatter) -> Result<(), ::core::fmt::Error> {
```
Comment (bkchr) [inline core/primitives/src/hexdisplay.rs]:
```suggestion
	fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
```

File: core/state-machine/src/ext.rs [hunk 11]
```diff
@@ -183,66 +187,137 @@ where	H: Hasher,
...
 
 	fn original_child_storage(&self, storage_key: ChildStorageKey, key: &[u8]) -> Option<Vec<u8>> {
 		let _guard = panic_handler::AbortGuard::force_abort();
-		self.backend.child_storage(storage_key.as_ref(), key).expect(EXT_NOT_ALLOWED_TO_FAIL)
+		let result = self.backend.child_storage(storage_key.as_ref(), key).expect(EXT_NOT_ALLOWED_TO_FAIL);
+		trace!(target: "state-trace", "{:04x}: ChildOriginal({}) {}={:?}",
+			self.id,
+			HexDisplay::from(&storage_key.as_ref()),
+			HexDisplay::from(&key),
+			result.as_ref().map(HexDisplay::from)
```
Comment (bkchr) [inline core/state-machine/src/ext.rs]:
```suggestion
			result.as_ref().map(HexDisplay::from),
```

File: core/state-machine/src/lib.rs [hunk 12]
```diff
@@ -258,6 +259,13 @@ impl<'a, B, H, N, T, O, Exec> StateMachine<'a, B, H, N, T, O, Exec> where
 			self.offchain_ext.as_mut().map(|x| &mut **x),
 			self.keystore.clone(),
 		);
+		let id = externalities.id;
+		trace!(target: "state-trace", "{:04x}: Call {} at {:?}. Input={:?}",
+			id,
+			self.method,
+			self.backend,
+			HexDisplay::from(&self.call_data)
```
Comment (bkchr) [inline core/state-machine/src/lib.rs]:
```suggestion
			HexDisplay::from(&self.call_data),
```

File: core/primitives/src/hexdisplay.rs (hunk 5 above)
Comment (arkpar) [reply core/primitives/src/hexdisplay.rs]: Don't see any good reason for dynamic dispatch here

File: core/client/db/src/lib.rs (hunk 6 above)
Comment (arkpar) [reply core/client/db/src/lib.rs]: Type name is not that important for debugging here

File: core/state-machine/src/ext.rs [hunk 13]
```diff
@@ -120,6 +123,7 @@ where
 			changes_trie_transaction: None,
 			offchain_externalities,
 			keystore,
+			id: rand::random(),
```
Comment (cheme) [inline core/state-machine/src/ext.rs]: Any ways to use a sequential index instead, or to lazyly init it behind a refcell?

Comment (cheme) [reply core/state-machine/src/ext.rs]: I can do a pr for it if it sounds usefull (refcell).

Comment (arkpar) [reply core/state-machine/src/ext.rs]: I thought about that but decided that random is good enough for debugging. Collision chance is rather small. For sequential counter I'd probably use `AtomicU16::fetch_add`

Comment (cheme) [reply core/state-machine/src/ext.rs]: Oh yes, atomic u16 seems better than rand to me for any consideration, and probably not worth lazy loading with a refcell :+1: 
(that was more the call to rand than the collision possibility that was bugging me).
